### ISSUE TITLE
Fused-compile training path InvalidCastException on T=double (Conv2D/VGG path)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#340]
## Repro

Tensors 0.77.1, AiDotNet 0.113.0 branch `fix/pr-1290-ci-failures`.
Running `VGGNetwork<double>` parameter-less ctor (224×224×3 → 1000 classes) through `NeuralNetworkBase<T>.TrainWithTape`, which routes through `CompiledTapeTrainingStep<T>.TryStepWithFusedOptimizer`:

```
ResNetPerfHarness Warning: 0 : CompiledTapeTrainingStep.TryStepWithFusedOptimizer
failed, falling back to eager:
System.InvalidCastException: Unable to cast object of type 'System.Double[]'
to type 'System.Single[]'.
   at AiDotNet.Training.CompiledTapeTrainingStep`1.TryStepWithFusedOptimizer(
        IReadOnlyList`1 layers, Tensor`1 input, Tensor`1 target,
        Func`2 forward, Func`3 computeLoss,
        OptimizerType optimizerType,
        Single learningRate, Single beta1, Single beta2, Single epsilon, Single weightDecay,
        T& lossValue)
    in C:\Users\cheat\source\repos\AiDotNet-pr1299\src\Training\CompiledTapeTrainingStep.cs:line 336
```

Line 336 is the actual `plan.Step()` invocation. The stack frame ends there, so the cast itself is upstream inside `CompiledTrainingPlan<T>.Step()`'s call chain.

`T = double`, default `AdamOptimizer<double>`, `UseAdaptiveLearningRate = false`, every other documented gate is passed (PR #322's `ConfigureOptimizer` correctly dispatches to `ConfigureOptimizerDouble` for `T == double`).

## Hypothesis

`CompiledTrainingPlan.cs` has several specialized-kernel paths that contain hardcoded `(float[])(object)tensor.GetDataArray()` casts:

- L316/L318 — gated by `T == float` (correct)
- L393/L395 — gated by `T == double` (correct)
- L793/L819 — Conv2D / matmul specialised paths
- L866/L875/L912 — SIMD ReLU backward etc.

The fused-double path lit up in PR #319 / #322 / #324 / #326 covers `ConfigureOptimizerDouble` + Conv2D forward & backward + BatchNorm4D backward, but at least one specialised backward / SIMD kernel along the VGG (Conv → BN → ReLU → Conv → ... → FlattenLayer → Dense × 3 → Softmax → CE-loss) trace still has a `(float[])(object)` cast that isn't gated on `T`.

## Why it matters

With v0.77.1's OpenBLAS-transitive fix and the perf work landed on the AiDotNet side, **fp64 eager is now 278 ms/step on VGG16-BN paper-scale** (down from 18.6 s, **67×**). That's fast enough to fit the model-family invariant suite's 120 s xUnit budget without the fused path — so this bug isn't blocking the immediate VGG CI failures. But:

- `Tensor<double>` is the default precision across the AiDotNet test base (`NeuralNetworkModelTestBase` is hardcoded on `INeuralNetworkModel<double>`).
- The fused-compile path is the production-fast path the rest of the AiDotNet codebase routes its CNN training through.
- Any user training a `<double>`-precision Conv-heavy model is silently falling back to eager. The fused-path comment block in `NeuralNetworkBase.cs:4842-4847` and `CompiledTapeTrainingStep.cs:231-235` both promise the double path works — currently they don't.

## Confirmed paths

- `fp32` `VGGNetwork<float>`: fused-compile **works**, 253 ms/step.
- `fp64` `VGGNetwork<double>`: cast exception, falls back to eager (278 ms/step on the new BLAS-engaged dispatch).

## Suggested investigation

Audit every `(float[])(object)` / `(Tensor<float>)(object)` cast in `CompiledTrainingPlan.cs` (and any sibling kernel files reached from `Step()`'s backward action chain) for missing `if (typeof(T) == typeof(double))` branches. The optimizer config wiring (PR #319 / #322) is already correct; the remaining mismatch is on specialised forward/backward SIMD paths invoked from inside `Step()`.

If a quick locate is helpful, the failing trace's last visible frame is line 336 in `AiDotNet/src/Training/CompiledTapeTrainingStep.cs` calling `plan.Step()` on a `<double>` plan with Adam configured.

Filed from AiDotNet PR #1299 investigation:
- AiDotNet branch: `fix/pr-1290-ci-failures` @ `eb16afaba`
- ResNetPerfHarness reproduces via: `dotnet run -- --model vgg16bn --dtype double --warmup 1 --iters 3`
